iroh-relay/src/client/conn.rs

```diff
@@ -3,280 +3,105 @@
-    defaults::timeouts::CLIENT_RECV_TIMEOUT,
-    protos::relay::{
-        write_frame, ClientInfo, Frame, RelayCodec, MAX_PACKET_SIZE, PER_CLIENT_READ_QUEUE_DEPTH,
-        PER_CLIENT_SEND_QUEUE_DEPTH, PROTOCOL_VERSION,
-    },
+    client::streams::MaybeTlsStreamChained,
+    protos::relay::{ClientInfo, Frame, RelayCodec, MAX_PACKET_SIZE, PROTOCOL_VERSION},
 };
 
-impl PartialEq for Conn {
-    fn eq(&self, other: &Self) -> bool {
-        Arc::ptr_eq(&self.inner, &other.inner)
-    }
+/// Error for sending messages to the relay server.
+#[derive(Debug, thiserror::Error)]
+pub enum ConnSendError {
+    /// An IO error.
+    #[error("IO error")]
+    Io(#[from] io::Error),
+    /// A protocol error.
+    #[error("Protocol error")]
+    Protocol(&'static str),
 }
 
-impl Eq for Conn {}
-
 /// A connection to a relay server.
 ///
-/// Cheaply clonable.
-/// Call `close` to shut down the write loop and read functionality.
-#[derive(Debug, Clone)]
-pub struct Conn {
-    inner: Arc<ConnTasks>,
-}
-
-/// The channel on which a relay connection sends received messages.
+/// This holds a connection to a relay server.  It is:
 ///
-/// The [`Conn`] to a relay is easily clonable but can only send DISCO messages to a relay
-/// server.  This is the counterpart which receives DISCO messages from the relay server for
-/// a connection.  It is not clonable.
-#[derive(Debug)]
-pub struct ConnReceiver {
-    /// The reader channel, receiving incoming messages.
-    reader_channel: mpsc::Receiver<Result<ReceivedMessage>>,
-}
-
-impl ConnReceiver {
-    /// Reads a messages from a relay server.
-    ///
-    /// Once it returns an error, the [`Conn`] is dead forever.
-    pub async fn recv(&mut self) -> Result<ReceivedMessage> {
-        let msg = self
-            .reader_channel
-            .recv()
-            .await
-            .ok_or(anyhow!("shut down"))??;
-        Ok(msg)
-    }
-}
-
+/// - A [`Stream`] for [`ReceivedMessage`] to receive from the server.
+/// - A [`Sink`] for [`SendMessage`] to send to the server.
+/// - A [`Sink`] for [`Frame`] to send to the server.
+///
+/// The [`Frame`] sink is a more internal interface, it allows performing the handshake.
+/// The [`SendMessage`] and [`ReceivedMessage`] are safer wrappers enforcing some protocol
+/// invariants.
 #[derive(derive_more::Debug)]
-pub struct ConnTasks {
-    /// Our local address, if known.
-    ///
-    /// Is `None` in tests or when using websockets (because we don't control connection establishment in browsers).
-    local_addr: Option<SocketAddr>,
-    /// Channel on which to communicate to the server. The associated [`mpsc::Receiver`] will close
-    /// if there is ever an error writing to the server.
-    writer_channel: mpsc::Sender<ConnWriterMessage>,
-    /// JoinHandle for the [`ConnWriter`] task
-    writer_task: AbortOnDropHandle<Result<()>>,
-    reader_task: AbortOnDropHandle<()>,
+pub(crate) enum Conn {
+    Relay {
```

Comment:
Same thing with the naming here
```suggestion
    Derp {
```